### DISCUSSION TITLE
Distribution vs package naming

### DISCUSSION THREAD

**Discussion by labdmitriy** [cjolowicz/cookiecutter-hypermodern-python#1200] [Q&A] [answered]
Hi @cjolowicz,

Thank you for your great series of articles and this template, this is very useful and I am learning a lot!

I have a question about naming for distributions and packages.

I saw the following statement in the [Chapter 1](https://cjolowicz.github.io/posts/hypermodern-python-01-setup/):
> Use [snake case](https://en.wikipedia.org/wiki/Snake_case) for the package name hypermodern_python, as opposed to the [kebab case](https://en.wiktionary.org/wiki/kebab_case) used for the repository name hypermodern-python. In other words, name the package after your repository, replacing hyphens by underscores.

Surprisingly for me I can't explain to myself why hyphens are used for distributions, not underscores, and didn't find any best practices on it.
Therefore I decided to write an [article](https://labdmitriy.github.io/blog/distributions-vs-packages/) with my thoughts about it, but still I don't understand why hyphens are so common for distribution names :smile:

Could you please tell what is the reason for you to choose hyphens for distribution names?

Thank you.

**Comment by cjolowicz** [comment, marked as answer]
Thanks for asking an interesting question, and for writing the article! I particularly liked that you gathered data about existing naming practices from PyPI.

Personally, I prefer package names with hyphens because I find them easier to read. Hyphens have been a part of standard typography [since Gutenberg](https://en.wikipedia.org/wiki/Hyphen#Origin_and_history).[^1]

[^1]: Interestingly, before Gutenberg, the hyphen looked more like an underscore. Gutenberg was forced to move the hyphen to the middle of the line for mechanical reasons.

By contrast, we've had the underscore character only since the advent of typewriters. According to [this thesis](https://dash.harvard.edu/bitstream/handle/1/33797379/ROMANO-DOCUMENT-2016.pdf), it was a fixture of the keyboard by 1881, and used for underlining by backing up the carriage and typing over the previous letter. It became a part of ASCII in 1963, and the C programming language made it a legal character in identifiers in the 1970s. I think you'd be hard-pressed to find a text with underscores that's not aimed at technical people. What's more, the original purpose of underscores was to provide underlining for devices that support inserting multiple characters in the same position. I still find it somewhat weird to read identifiers that underline the gaps between words, but not the words themselves.

This is obviously a personal choice, and there are well-known projects using underscores in their names. To name one prominent example, the import sorter previously used by this template was [reorder_python_imports](https://github.com/asottile/reorder_python_imports). Using underscores for both distribution and import names has the advantage of consistency, as hyphens are not legal characters in Python identifiers, while underscores are.

Worth noting that [PEP 503](https://peps.python.org/pep-0503/#normalized-names) treats the three non-alphanumeric characters (`.`, `-`, `_`) in package names as equivalent to `-`. So the above example appears as [reorder-python-imports](https://pypi.org/project/reorder-python-imports/) on PyPI.

**Comment by labdmitriy** [comment]
Wow thank you for such comprehensive answer!

Would you mind letting me include this information to the update of my blog post (referring to your answer)?

I definitely need to restructure and enhance the description in the post:
- I want to consider the case where the repository name differs more than by hyphen or underscore. 
  The example of such package I saw recently is [tzwhere](https://github.com/pegler/pytzwhere))
- Do I understand correclty that even when we specify distribution name in setup.py/setup.cfg or pyproject.toml, this name will be normalized for the name on PyPI following the rules you mentioned? I didn't know about that.
  Package `reorder_python_imports` is an excellent example.  
  But as I understand from the naming stats from PyPI, there are distributions with underscores only or even hyphens at the same time, probably these distributions were created before normalization was implemented?

Thank you.

**Reply by cjolowicz** [reply]
> Would you mind letting me include this information to the update of my blog post (referring to your answer)?

Sure, you're welcome!

> I want to consider the case where the repository name differs more than by hyphen or underscore.

This is quite common, especially the case with a `python-` or `py` prefix:

- python-dateutil
- pyyaml

> Do I understand correctly that even when we specify distribution name in setup.py/setup.cfg or pyproject.toml, this name will be normalized for the name on PyPI following the rules you mentioned?

PyPI preserves the original name in the package metadata, as specified by the package author. The URL uses the normalized name ([PEP 503](https://peps.python.org/pep-0503/#normalized-names)), but name variants are accepted and redirected to the normalized name.

PEP 503 normalization also includes transforming to lowercase. For example, the metadata name for pyyaml is `PyYAML`, whose canonical form is `pyyaml`.

There is also the question of normalizing distribution filenames. For wheels, this is specified in [PEP 427](https://peps.python.org/pep-0427/#escaping-and-unicode) and based on underscores and original case (for pyyaml, the wheel filename uses `PyYAML`). For sdists, this is unspecified (see [this blog](https://blog.yossarian.net/2022/05/09/A-most-vexing-parse-but-for-Python-packaging) and [this issue](https://github.com/pypa/packaging/issues/527)).

Tools like pip or Poetry will work with any name variant. For example, Poetry uses the original name from the package metadata (`PyYAML`) in pyproject.toml and poetry.lock, but you can `poetry add pyyaml`.

---

As an aside, name variants also exist for the console script (where applicable).

- cogapp (cog)
- rst-to-myst (rst2myst)

Sometimes, only the "marketing name" and/or the repository name are different:

- coverage (distribution, package, and script are named `coverage`, repository is named `coveragepy`, human-friendly name is `Coverage.py`)

**Reply by labdmitriy** [reply]
Thanks a lot for this detailed explanation! 
This is more tricky than I expected :sweat_smile: 

Maybe this question is not related to Python directly, but your answer is an interesting historical note itself :smile: 

I will read your links and will update the post!

**Comment by labdmitriy** [comment]
> PyPI preserves the original name in the package metadata, as specified by the package author. The URL uses the normalized name ([PEP 503](https://peps.python.org/pep-0503/#normalized-names)), but name variants are accepted and redirected to the normalized name.

I decided to make a small experiment - get the latest by update time package with both underscores and hyphens and now it is [pyprocessors-mazars_table](https://pypi.org/project/pyprocessors-mazars_table/) and it seems that URL is redirected from different URL variants to the **original** one, not **normalized** version of the name.

There are several examples of the redirections:
- https://pypi.org/project/pyprocessors-mazars_table/ -> https://pypi.org/project/pyprocessors-mazars_table/
- https://pypi.org/project/pyprocessors-mazars-table/ -> https://pypi.org/project/pyprocessors-mazars_table/
- https://pypi.org/project/pyprocessors-mazars.table/ -> https://pypi.org/project/pyprocessors-mazars_table/
- https://pypi.org/project/pyprocessors.mazars_table/ -> https://pypi.org/project/pyprocessors-mazars_table/ 
- https://pypi.org/project/pyprocessors_mazars_table/ -> https://pypi.org/project/pyprocessors-mazars_table/

**Reply by cjolowicz** [reply]
That's an interesting find! The topic of name normalization comes up regularly on the Python Discourse, in the [Packaging category](https://discuss.python.org/c/packaging). This is a place where both the status quo and the future standards of the Python packaging ecosystem are discussed, and it's well worth following and engaging with.

**Reply by labdmitriy** [reply]
Thank you for the link!

**Comment by labdmitriy** [comment]
Hi Claudio,

I substantially restructured [the article](https://labdmitriy.github.io/blog/distributions-vs-packages/) and also added the information that you provided.

Thanks a lot for your help!
